source/agora/test/ValidatorCleanRestart.d

```diff
@@ -0,0 +1,150 @@
+        .filter!(tx => tx.type == TxType.Payment)
+        .map!(tx => iota(tx.outputs.length)
+            .map!(idx => TxBuilder(tx, cast(uint)idx)))
+        .joiner().array;
+
+    // Discarded UTXOs (just to trigger block creation)
+    auto txs = spendable[0 .. 5]
+        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign())
+        .array;
+
+    // 8 utxos for freezing, 8 utxos for creating a block later
+    txs ~= spendable[5].split(WK.Keys.byRange.take(8).map!(k => k.address)).sign();
+    txs ~= spendable[6].split(WK.Keys.Genesis.address.repeat(8)).sign();
+    txs ~= spendable[7].split(WK.Keys.Genesis.address.repeat(8)).sign();
+
+    // Block 8
+    txs.each!(tx => set_a[0].putTransaction(tx));
+    network.expectBlock(Height(8), 5.seconds);
+
+    // Freeze builders
+    auto freezable = txs[$ - 3]
+        .outputs.length.iota
+        .takeExactly(8)
+        .map!(idx => TxBuilder(txs[$ - 3], cast(uint)idx))
+        .array;
+
+    // Create 8 freeze TXs
+    auto freeze_txs = freezable
+        .enumerate
+        .map!(pair => pair.value.refund(WK.Keys[pair.index].address)
+            .sign(TxType.Freeze))
+        .array;
+    assert(freeze_txs.length == 8);
+
+    // Block 9
+    freeze_txs.each!(tx => set_a[0].putTransaction(tx));
+    network.expectBlock(Height(9), 5.seconds);
+
+
+    // Now we enroll 4 new validators. After this, the already enrolled
+    // validators will be expired.
+    foreach (ref node; set_b)
+    {
+        Enrollment enroll = node.createEnrollmentData();
+        node.enrollValidator(enroll);
+
+        // Check enrollment
+        set_b.each!(node =>
+            retryFor(node.getEnrollment(enroll.utxo_key) == enroll, 5.seconds));
+    }
+
+    // Block 10
+    auto new_txs = txs[$ - 2]
+        .outputs.length.iota.map!(idx => TxBuilder(txs[$ - 2], cast(uint)idx))
+        .takeExactly(8)
+        .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
+    new_txs.each!(tx => set_a[0].putTransaction(tx));
+    network.expectBlock(Height(10), 5.seconds);
+
+    auto b10 = set_a[0].getBlocksFrom(0, 2)[0];
+    assert(b10.header.enrollments.length == 4);
+
+    // Now restarting the validators in the set B, all the date of those
+    // validators has been wiped out.
+    set_b.each!(node => network.restart(node));
+    network.expectBlock(Height(10), 8.seconds);
+
+    // Sanity check
+    nodes.enumerate.each!((idx, node) =>
+        retryFor(node.getValidatorCount == 4, 3.seconds,
+            format("Node %s has validator count %s. Expected: %s",
+                idx, node.getValidatorCount(), 4)));
+
+    // Check the connection state is complete for the set B
+    set_b.each!(node =>
+        retryFor(node.getNodeInfo().state == NetworkState.Complete, 8.seconds));
+
+    foreach (node; set_b)
+        writeln("Addresses: ", node.getNodeInfo().addresses);
+
+    // Shutdown all the validators of the set A
+    // set_a.each!(node => node.shutdown());
```

Comment:
If this doesn't work you can replace it with `node.ctrl.sleep(time, true)`. This will make the node ignore all requests - which is a similar effect to shutting it down.